### ISSUE TITLE
Move CUDA into an extension

### ISSUE THREAD

**Issue by navidcy** [CliMA/Oceananigans.jl#3481]
It'd be nice if we move CUDA functionality into an extension so it's only loaded for CUDA-enabled devices.

@vchuravy made a first attempt to do that at https://github.com/CliMA/Oceananigans.jl/pull/3066

similar effort but for AMDGPU is at https://github.com/CliMA/Oceananigans.jl/pull/3475

**Comment by navidcy** [comment]
Of course, there is a (minor?) downside to this. As things stand now we can do:

```Julia
using Oceananigans

grid = RectilinearGrid(GPU(), ...)

...
```

If we move CUDA functionality to an extension we'd have to do:

Of course, there is a (minor?) downside to this. As things stand now we can do:

```Julia
using Oceananigans
using CUDA

grid = RectilinearGrid(GPU(), ...)

...
```

**Comment by vchuravy** [comment]
The benefit is that the user can choose to not install CUDA and just use the CPU implementation of Oceananigans, or use AMDGPU instead of CUDA. So installing Oceanigans will be faster and users on laptops without GPUs won't have to pay for the additional packages

**Comment by glwagner** [comment]
I think the benefit of faster CPU compilation is worth the price of a `using CUDA`, not least because we should be concerned about time-to-first-solution. It also seems more forward-looking since there may be many GPU options in the near future.

**Comment by ffrancesco94** [comment]
I guess this fell through the cracks somehow..?

**Comment by navidcy** [comment]
Lack of time mostly. Not lack of interest. But if someone leads a PR I can try help!

**Comment by ffrancesco94** [comment]
I'm not super familiar with the extension system (from the developer side) but let me dig around the documentation and see what I can come up with it!

**Comment by glwagner** [comment]
There is already a PR open which is very close to completion: https://github.com/CliMA/Oceananigans.jl/pull/4499
